### PR TITLE
Circumvent URL Injections in various Cogs

### LINKED ISSUE #801: URL Injection
# Explanation

It's like SQL injection, but for URLs. Carelessly formatting user input into a URL template can allow attackers construct a specific string to do something nasty. No direct exploit has been found, but the practices used are risky

# To fix

https://github.com/python-discord/sir-lancebot/blob/main/bot/exts/evergreen/movie.py
URL formatting partially using f-strings (line 145). I wasn't able to find a vulnerability, but no check is made to ensure that the substitution is safe

https://github.com/python-discord/sir-lancebot/blob/main/bot/exts/evergreen/wikipedia.py
Vulnerable, try `.wiki JavaScript` vs `.wiki JavaScript&srlimit=1`.

https://github.com/python-discord/sir-lancebot/blob/main/bot/exts/evergreen/githubinfo.py
Vulnerable, bot fails on `.github user gvanrossum/followers`.

https://github.com/python-discord/sir-lancebot/blob/main/bot/exts/halloween/hacktoberstats.py
URL formatting using concatenation. I don't see a direct vulnerability, but the cog does use a GitHub token, so a check should be added to be sure

# Who's going to fix it

Feel free to address this. If nobody does, I'll be able to assign myself in about a week

### PR DESCRIPTION
**Pull request by D0rs4n**
## Relevant Issues

Closes  #801

## Description

I applied the built in tools from aiohttp wherever I could to properly sanitize the inputs, in case of the GithubInfo Cog, because of the API architecture I had to use a different approach, but by switching to `quote_plus` I managed to fix the issue.
The cogs fixed:
- Movie
- Githubinfo
- Wikipedia
- Hacktoberstats (get_october_prs function in particular [link](https://github.com/python-discord/sir-lancebot/blob/main/bot/exts/halloween/hacktoberstats.py#L217-L228))

## Did you:

- [x] Join the [**Python Discord Community**](https://discord.gg/python)?
- [x] Read all the comments in this template?
- [x] Ensure there is an issue open, or link relevant discord discussions?
- [x] Read and agree to the [contributing guidelines](https://pythondiscord.com/pages/contributing/contributing-guidelines/)?

### CODE REVIEW

File: bot/exts/evergreen/wikipedia.py [hunk 1]
```diff
@@ -35,8 +44,8 @@ def __init__(self, bot: Bot):
 
     async def wiki_request(self, channel: TextChannel, search: str) -> Optional[List[str]]:
         """Search wikipedia search string and return formatted first 10 pages found."""
-        url = SEARCH_API.format(number_of_results=10, string=search)
-        async with self.bot.http_session.get(url=url) as resp:
+        params = dict(WIKI_PARAMS, srlimit=10, srsearch=search)
```
Comment (decorator-factory) [inline bot/exts/evergreen/wikipedia.py:47]: I would use a literal: `{**WIKI_PARAMS, "srlimit": 10, "srsearch": search}` or `WIKI_PARAMS | {"srlimit": 10, "srsearch": search}`

File: bot/exts/evergreen/movie.py [hunk 2]
```diff
@@ -142,9 +141,9 @@ async def get_pages(self, client: ClientSession, movies: Dict[str, Any], amount:
 
     async def get_movie(self, client: ClientSession, movie: int) -> Dict:
         """Get Movie by movie ID from TMDB. Return result dictionary."""
-        url = BASE_URL + f"movie/{movie}?" + urlencode(MOVIE_PARAMS)
+        url = BASE_URL + f"movie/{movie}"
```
Comment (decorator-factory) [inline bot/exts/evergreen/movie.py:144]: I think you should add a check that `movie` is actually an integer, either here or after line 135 where it's fetched, but not validated

File: bot/exts/halloween/hacktoberstats.py [hunk 3]
```diff
@@ -208,24 +209,24 @@ async def get_october_prs(self, github_username: str) -> Optional[List[dict]]:
...
             f"+-is:{not_query}"
             f"+created:{date_range}"
             f"&per_page={per_page}"
         )
-        log.debug(f"GitHub query URL generated: {query_url}")
 
-        jsonresp = await self._fetch_url(query_url, REQUEST_HEADERS)
+        log.debug(f"GitHub query parameters generated: {query_params}")
+
+        jsonresp = await self._fetch_url(base_url, REQUEST_HEADERS, dict(q=query_params))
```
Comment (decorator-factory) [inline bot/exts/halloween/hacktoberstats.py:229]: I would do `{"q": query_params}`

**Review by decorator-factory — CHANGES_REQUESTED** [review]

**Review by decorator-factory — APPROVED** [review]
LGTM

File: bot/exts/evergreen/githubinfo.py [hunk 4]
```diff
@@ -91,7 +91,8 @@ async def github_user_info(self, ctx: commands.Context, username: str) -> None:
             )
 
             if user_data["type"] == "User":
-                embed.add_field(name="Gists", value=f"[{gists}](https://gist.github.com/{quote(username, safe='')})")
+                embed.add_field(name="Gists", value=f"[{gists}](https://gist.github.com/{quote_plus(username, safe='')}"
+                                                    f")")
```
Comment (MarkKoz) [inline bot/exts/evergreen/githubinfo.py:95]: The indentation here looks messed up.

**Review by Bluenix2 — APPROVED** [review]
This looks good, thank you for the PR

File: bot/exts/evergreen/githubinfo.py (hunk 4 above)
Comment (Bluenix2) [reply bot/exts/evergreen/githubinfo.py:95]: That's a visual indent.

That said, I do guess you could do a hanging indent like this:
```suggestion
                embed.add_field(
                    name="Gists",
                    value=f"[{gists}](https://gist.github.com/{quote_plus(username, safe='')})"
                )
```

Comment (D0rs4n) [reply bot/exts/evergreen/githubinfo.py:95]: Fair point, I wanted to pass the linting. I'll push a change shortly.

Comment (decorator-factory) [reply bot/exts/evergreen/githubinfo.py]: oh yeah I missed that, looks like something an autoformatter would do :eyes:

Comment (MarkKoz) [reply bot/exts/evergreen/githubinfo.py]: What Bluenix showed is the way it should be done. This is what's specified in our style guide, but I am not sure how strict this project is with enforcing that. In any case, I'm not going to make you push a second commit.

Comment (Bluenix2) [reply bot/exts/evergreen/githubinfo.py]: Yeah I agree with Mark, I prefer the hanging indent over visual

Comment (D0rs4n) [reply bot/exts/evergreen/githubinfo.py]: I can push another commit, it's not a problem! :)

File: bot/exts/evergreen/githubinfo.py [hunk 5]
```diff
@@ -91,7 +91,10 @@ async def github_user_info(self, ctx: commands.Context, username: str) -> None:
             )
 
             if user_data["type"] == "User":
-                embed.add_field(name="Gists", value=f"[{gists}](https://gist.github.com/{quote(username, safe='')})")
+                embed.add_field(
+                    name="Gists",
+                    value=f"[{gists}](https://gist.github.com/"f"{quote_plus(username, safe='')}"f")"
+                )
```
Comment (Bluenix2) [inline bot/exts/evergreen/githubinfo.py:97]:
```suggestion
                embed.add_field(
                    name="Gists",
                    value=f"[{gists}](https://gist.github.com/"f"{quote_plus(username, safe='')})"
                )
```

**Review by Bluenix2 — CHANGES_REQUESTED** [review]

**Comment by MarkKoz** [comment]
I'm not very familiar with URL injections, but you may have missed some:

There are three different places in the wolfram cog; this is just one of them
https://github.com/python-discord/sir-lancebot/blob/32b6d0e386350ea12ab9e745e2bede844d289d07/bot/exts/evergreen/wolfram.py#L111-L122

---

This one might be okay since there's no user input, but it could still be passed via the params kwarg
https://github.com/python-discord/sir-lancebot/blob/32b6d0e386350ea12ab9e745e2bede844d289d07/bot/exts/valentines/movie_generator.py#L38-L39

I think this is fine but it could've been simpler to use param kwargs instead
https://github.com/python-discord/sir-lancebot/blob/32b6d0e386350ea12ab9e745e2bede844d289d07/bot/exts/evergreen/stackoverflow.py#L35-L37

**Comment by D0rs4n** [comment]
I see, as for the Wolfram, I was told that it was out of the scope of this issue, but I'm happy to push patches to all of those.

**Comment by MarkKoz** [comment]
Does anyone know why it is out of scope? I wasn't involved in that discussion.

**Comment by D0rs4n** [comment]
Because, there were other kind of sanitization measures applied.
In case of wolfram that was `parse.urlencode`.

**Comment by decorator-factory** [comment]
@D0rs4n
Feel free to fix these consistency issues if you want to

**Comment by MarkKoz** [comment]
> Because, there were other kind of sanitization measures applied.
> In case of wolfram that was `parse.urlencode`.

Well, that was also the case with the movie cog but you made changes anyway.

**Comment by D0rs4n** [comment]
> > Because, there were other kind of sanitization measures applied.
> > In case of wolfram that was `parse.urlencode`.
> 
> Well, that was also the case with the movie cog but you made changes anyway.

Fair point, I pushed some changes to improve consistency! Hopefully everything's going to be fine! :)

**Review by decorator-factory — APPROVED** [review]

**Review by Bluenix2 — APPROVED** [review]
Yeah this looks good now, thank you.

**Comment by decorator-factory** [comment]
Thanks for the PR, crisis overted

File: bot/exts/evergreen/githubinfo.py [hunk 6]
```diff
@@ -120,7 +123,7 @@ async def github_repo_info(self, ctx: commands.Context, *repo: str) -> None:
             return
 
         async with ctx.typing():
-            repo_data = await self.fetch_data(f"{GITHUB_API_URL}/repos/{quote(repo)}")
+            repo_data = await self.fetch_data(f"{GITHUB_API_URL}/repos/{quote_plus(repo)}")
```
Comment (MarkKoz) [inline bot/exts/evergreen/githubinfo.py:126]: This somehow broke the command. It fails to find the repo with `.gh repo python/cpython`, but that works fine on the main branch.

**Review by MarkKoz — CHANGES_REQUESTED** [review]
I can't really test hacktoberstats since the event isn't going on right now.

File: bot/exts/evergreen/githubinfo.py (hunk 6 above)
Comment (D0rs4n) [reply bot/exts/evergreen/githubinfo.py:126]: Yeah, I should have only used it when we are looking up usernames, this piece of code right now encodes the `/` sign.

**Review by MarkKoz — APPROVED** [review]

**Pull request merged**